start-loggerhead.1
.TH START-LOGGERHEAD "1" "July 2008" "start-loggerhead 1.2.1" "User Commands"
.SH NAME
start-loggerhead \- start a loggerhead server
.SH SYNOPSIS
.B start-loggerhead
[\fIoptions\fR]
.SH DESCRIPTION
Loggerhead is a web viewer for projects in bazaar. It can be used to navigate 
a branch history, annotate files, view patches, perform searches, etc. It's 
heavily based on bazaar-webserve, which is itself based on hgweb for Mercurial.
.PP
The start-loggerhead command starts a new standalone loggerhead server.
.SH OPTIONS
.TP
\fB\-\-version\fR
show program's version number and exit
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-f\fR, \fB\-\-foreground\fR
run in the foreground; don't daemonize
.TP
\fB\-C\fR, \fB\-\-check\fR
only start if not already running (useful for cron jobs)
.TP
\fB\-p\fR, \fB\-\-pidfile=PIDFILE\fR
override pid file location
.TP
\fB\-c\fR, \fB\-\-config-file=CONFIGFILE\fR
override configuration file location
.TP
\fB\-L\fR, \fB\-\-log-folder=LOG-FOLDER\fR
override log directory
.SH "LICENSE"
GNU GPLv2 or later.
.SH "SEE ALSO"
https://launchpad.net/loggerhead


NEWS
What's changed in loggerhead?
=============================

dev [future]
------------

1.18 [10Nov2010]
----------------

    - Remove ``start-loggerhead`` and ``stop-loggerhead`` which were already
      deprecated. (John Arbash Meinel)

    - Syntax highlighting is no longer applied for files greater than 512K,
      reducing codebrowse.launchpad.net overloading.
      (Max Kanat-Alexander, #513044)

    - Documentation added in the docs directory. README simplified
      accordingly. (Tres Seaver).

    - Show svn/git/hg revision ids in loggerhead revision view.
      (Jelmer Vernooij)

    - Fix .bzr/smart access to branches in shared repos. (You also need
      a version of bzr with bug #348308 fixed.) (Andrew Bennetts)

    - Support FastCGI, SCGI and AJP using flup. (Denis Martinez)

    - Repository.get_revision_inventory() was removed in bzr 2.2; use
      Repository.get_inventory() instead. (Matt Nordhoff, #528194)

    - Ignore readonly+ prefix when checking if Loggerhead is serving a
      local location. (Reported by Tres Seaver.) (Matt Nordhoff)

    - Set Cache-Control and Expires headers on static pages.
      (John Arbash Meinel)

    - Generate relative links where possible (everywhere but HTTP
      redirects and feed IDs). (Michael Hudson, Matt Nordhoff)

    - Fix bad redirect when visiting "/download" or "/download/".
      (Matt Nordhoff, #247992)


1.17 [20Aug2009]
----------------

    - Add bug links in revision informations (Alexandre Garnier, #314052)

    - Make sure that binary files aren't annotated. (Martin Albisetti,
      #258848)

    - Loggerhead now serves bzr branches over HTTP and exposes the URL
      to branch them. Addresses bug #240577. (Jonathan Lange)

    - Leading blank lines in commit messages no longer result in an
      empty summary. (Colin Watson)

    - Added optional syntax highlighting to annotate view using
      python-pygments.  Partially addresses bug #306631. (Peter Bui)

    - Convert newlines in commit messages to HTML line breaks for
      annotate and changelog views.  Addresses bug #273688. (Peter
      Bui)

    - serve-branches now errors if run behind a proxy without
      paste.deploy installed. (Michael Hudson)

    - Loggerhead should now handle file and directory names that need
      URL escaping without crashing.

    - The start-loggerhead script properly sets the wsgi.url_scheme
      from the server.webpath option. (neror, #260547)

    - The revision page defaults to unified style again, and can
      convert to a side-by-side view using JavaScript. (Michael Hudson)

    - Clean up and improve performance of the annotate view. (Michael
      Hudson)

    - Finish converting JavaScript from MooTools to YUI 3. (Michael
      Hudson)

    - Improve compatibility with IE 6. (Michael Hudson)

    - Leading blank lines in commit messages no longer result in an
      empty summary. (Colin Watson)

    - Clip long lines in side-by-side diff view. (Michael Hudson,
      #334837)

    - The user-confusing "next" and "previous" links now read "older"
      and "newer" respectively. (Michael Hudson, #297930)

    - The annotate view now contains line number anchors. (Michael
      Hudson)

    - Fix inventory pages using "//" in links. (Michael Hudson, #329668)

    - Fix problems viewing files and directories containing spaces and
      other funny characters. (Peter Bui)

    - Changelog messages are now displayed with newlines preserved.
      (Peter Bui, #273688)

    - Offer a link to see the full file diffs for a file path. (Michael
      Hudson, #333797)

    - Fix annotate error caused by Pygments stripping trailing
      whitespace. (Michael Hudson, #338762)

    - Loggerhead can be installed as a Bazaar plugin and run by
      'bzr serve --http'. (Martin Pool)

    - Load parts of the changelog and revision pages via XMLHttpRequest
      to improve performance. This adds a dependency on simplejson or
      json. Partially addresses bug #253950. (Michael Hudson)

    - Various improvements to the animation JavaScript. (Michael Hudson)

    - Fix HTML content of source files being displayed unescaped when
      Pygments was unavailable. (Michael Hudson, #344970)

    - Fix serve-branches's path argument. (Michael Hudson, #353230)

    - serve-branches now has an option, --use-cdn, to load YUI from
      Yahoo!'s CDN. (Matt Nordhoff)

    - Fix certain race conditions for loading bzr-search. (Robert
      Collins, #334250)

    - Fix errors when using serve-branches --log-folder or --user-dirs.
      (It was calling config.get_option() incorrectly.) (Matt Nordhoff,
      bug #361238)

    - Move some caching from RAM to the disk, and other caching and
      memory usage improvements. (Michael Hudson)

    - Add a --cache-dir option to serve-branches to choose where to
      place the SQL cache, and only create one temporary SQL dir per
      process. (Matt Nordhoff, #358322)

    - Replace homebrew memory profiling code with Dozer. (Paul Hummer)

    - Use the branch's public_branch as the default suggested URL to
      branch from (Matt Nordhoff, #369767)

    - Fix a file descriptor leak (Matt Nordhoff, #370845)

    - Use transport API internally, so it is possible to specify a remote
      URL to serve-branches. (Jelmer Vernooij, #371787)

    - Fix internal server errors when using start-loggerhead. (Matt
      Nordhoff, #375948)

    - Fix annotating non-UTF-8 files when Pygments is disabled. (Matt
      Nordhoff, #376957)

    - Fix 'bzr serve --http' errors. (Matt Nordhoff, #377551)

    - Added the option to hide branches by setting http_serve = False
      in locations.conf (Martin Albisetti)

    - Fix serving branches over HTTP. (Matt Nordhoff, Jelmer Vernooij,
      #380026)

    - Install loggerhead as a bzr plugin by default (Jelmer Vernooij)

    - Fix logging 404 Not Found responses (Matt Nordhoff, #381029)

    - Bumped minimunm bzrlib version to 1.13 (Martin Albisetti)


start-loggerhead.py
#!/usr/bin/env python

import logging
import logging.handlers
from optparse import OptionParser
import os
import sys
import urlparse

from configobj import ConfigObj

from paste import httpserver
from paste.httpexceptions import make_middleware
from paste.translogger import make_filter

from loggerhead import daemon, release
from loggerhead.apps.config import Root


def make_handler(config, filename):
    roll = config.get('log.roll', 'never')
    if roll == 'daily':
        h = logging.handlers.TimedRotatingFileHandler(filename, 'midnight', 0, 100)
    elif roll == 'weekly':
        h = logging.handlers.TimedRotatingFileHandler(filename, 'W0', 0, 100)
    else:
        h = logging.FileHandler(filename)
    return h


def setup_logging(home, config, foreground):
    # i hate that stupid logging config format, so just set up logging here.

    log_folder = os.path.join(home, 'logs')
    if not os.path.exists(log_folder):
        os.mkdir(log_folder)

    f = logging.Formatter('%(levelname)-.3s [%(asctime)s.%(msecs)03d] %(name)s: %(message)s',
                          '%Y%m%d-%H:%M:%S')
    debug_log = make_handler(config, os.path.join(log_folder, 'debug.log'))
    debug_log.setLevel(logging.DEBUG)
    debug_log.setFormatter(f)
    if foreground:
        stdout_log = logging.StreamHandler(sys.stdout)
        stdout_log.setLevel(logging.DEBUG)
        stdout_log.setFormatter(f)
    f = logging.Formatter('[%(asctime)s.%(msecs)03d] %(message)s',
                          '%Y%m%d-%H:%M:%S')
    access_log = make_handler(config, os.path.join(log_folder, 'access.log'))
    access_log.setLevel(logging.INFO)
    access_log.setFormatter(f)

    logging.getLogger('').addHandler(debug_log)
    logging.getLogger('turbogears.access').addHandler(access_log)
    logging.getLogger('turbogears.controllers').setLevel(logging.INFO)

    if foreground:
        logging.getLogger('').addHandler(stdout_log)



def main():
    parser = OptionParser(usage='usage: %prog [options]', version='%prog ' + release.version)
    parser.add_option('-f', '--foreground', action='store_true', dest='foreground', default=False,
                      help="run in the foreground; don't daemonize")
    parser.add_option('-C', '--check', action='store_true', dest='check', default=False,
                      help="only start if not already running (useful for cron jobs)")
    options, args = parser.parse_args()
    if len(args) > 0:
        parser.error('No filename arguments are used, only options.')


    home = os.path.realpath(os.path.dirname(__file__))
    pidfile = os.path.join(home, 'loggerhead.pid')

    if options.check:
        if daemon.is_running(pidfile):
            sys.exit(0)
        sys.stderr.write('Did not find loggerhead running in %r; restarting...\n' % (pidfile,))

    # read loggerhead config

    config = ConfigObj(os.path.join(home, 'loggerhead.conf'), encoding='utf-8')
    extra_path = config.get('bzrpath', None)
    if extra_path:
        sys.path.insert(0, extra_path)

    #turbogears.update_config(configfile="dev.cfg", modulename="loggerhead.config")

    potential_overrides = [ ('server.socket_port', int),
                            ('server.webpath', str),
                            ('server.thread_pool', int),
                            ('server.socket_host' ,str) ]
    server_port = int(config.get('server.socket_port', 8080))
    nworkers = int(config.get('server.thread_pool', 10))
    server_host = config.get('server.socket_host', '0.0.0.0')
    webpath = config.get('server.webpath', None)

    for key, keytype in potential_overrides:
        value = config.get(key, None)
        if value is not None:
            value = keytype(value)
            #turbogears.config.update({ key: value })

    if not options.foreground:
        sys.stderr.write('\n')
        sys.stderr.write('Launching loggerhead into the background.\n')
        sys.stderr.write('PID file: %s\n' % (pidfile,))
        sys.stderr.write('\n')

        daemon.daemonize(pidfile, home)

    setup_logging(home, config, foreground=options.foreground)

    log = logging.getLogger('loggerhead')
    log.info('Starting up...')

    app = Root(config)

    app = app
    app = make_middleware(app)
    app = make_filter(app, None)

    if webpath:
        scheme, netloc, path, blah, blah, blah = urlparse.urlparse(webpath)
        def app(environ, start_response, orig=app):
            environ['SCRIPT_NAME'] = path
            environ['HTTP_HOST'] = netloc
            return orig(environ, start_response)

    try:
        httpserver.serve(
            app, host=server_host, port=server_port,
            threadpool_workers=nworkers)
    finally:
        log.info('Shutdown.')
        try:
            os.remove(pidfile)
        except OSError:
            pass


if __name__ == '__main__':
    main()
